File: Backend/src/test/java/lan/dk/podcastserver/utils/custom/ffmpeg/ProcessListenerTest.java
package lan.dk.podcastserver.utils.custom.ffmpeg;

import io.vavr.control.Try;
import org.junit.Test;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

import static com.jayway.awaitility.Awaitility.await;
import static java.util.concurrent.CompletableFuture.runAsync;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;

/**
 * Created by kevin on 24/07/2016.
 */
public class ProcessListenerTest {

    @Test
    public void should_wait_if_no_process() {
        /* Given */
        ProcessListener pl = new ProcessListener("foo");
        Process aProcess = mock(Process.class);
        final Process[] process = {null};

        /* When */
        runAsync(() -> {
            try { process[0] = pl.getProcess().get(); } catch (InterruptedException | ExecutionException ignored) {}
        });
        Try.run(() -> TimeUnit.MILLISECONDS.sleep(200));
        pl.process(aProcess);

        /* Then */
        await().atMost(5, TimeUnit.SECONDS).until(() -> {
            assertThat(process[0]).isNotNull();
        });
    }

}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/utils/custom/ffmpeg/ProcessListenerTest.kt
package com.github.davinkevin.podcastserver.utils.custom.ffmpeg

import com.jayway.awaitility.Awaitility.await
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.mockito.Mockito
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit.MILLISECONDS
import java.util.concurrent.TimeUnit.SECONDS

internal class ProcessListenerTest {

    @Test
    fun `should wait if no process until one is present`() {
        /* Given */
        val pl = ProcessListener("foo")
        val aProcess = Mockito.mock(Process::class.java)
        var process: Process? = null

        /* When */
        CompletableFuture.runAsync { process = pl.process.get() }
        MILLISECONDS.sleep(200)
        pl.withProcess(aProcess)

        /* Then */
        await().atMost(5, SECONDS).until {
            assertThat(process).isNotNull().isSameAs(aProcess)
        }
    }

}